### DISCUSSION TITLE
LVGL - How to correctly read .bin images from the SD card? (Or memory leak bug)

### DISCUSSION THREAD

**Discussion by svyat-innowise** [zephyrproject-rtos/zephyr#73072] [Q&A] [answered]
Hello everyone,

I encountered a possible memory leak in LVGL.

I want to retrieve a .bin image from an SD card to display it on the screen. Before that, I converted the image to .bin format using https://lvgl.io/tools/imageconverter. The image reads and displays well on the screen.

BUT I want to switch between different displays, removing the old display and getting a new one, which also has an image that I want to load from the SD card. Upon re-reading the image, even after clearing the cache, memory leaks. However, without loading the image from the SD card, the memory is released correctly.

To reproduce the situation, I use the following code:

```
while (true) {
        lv_obj_t * old_src = lv_scr_act();
        lv_img_cache_invalidate_src(NULL);

        lv_obj_t * current_screen = lv_obj_create(NULL);
        lv_obj_t * image = lv_img_create(current_screen);
        lv_img_set_src(image, "/SD:/res/ui/image.bin");

        lv_scr_load(current_screen);
        lv_obj_del(old_src);

        lv_task_handler();
        k_sleep(K_MSEC(100));
        LOG_INF("SCREEN with image has been loaded");
        k_sleep(K_SECONDS(1));
}
```
![image](https://github.com/zephyrproject-rtos/zephyr/assets/142225488/611cca35-7430-4426-9ed7-1bc3c520bc70)

An example of the resulting debug is attached below in the image:

Can anyone suggest a solution to this issue? Or maybe I am not freeing image resources correctly in LVGL? I would appreciate any help.

I am using:
Zephyr RTOS v3.5.0 https://github.com/zephyrproject-rtos/zephyr/releases/tag/v3.5.0
LVGL v8.4.0 https://github.com/zephyrproject-rtos/lvgl with default lvgl memory allocation functions

**Comment by svyat-innowise** [comment, marked as answer]
Solved. Problem was in the lvgl_fs. file. Memory wasn't released in the lvgl_fs_close function. https://devzone.nordicsemi.com/f/nordic-q-a/108039/nrf52840dk-lvgl-file-system

Patch for fix it:
```
From c6d24f3fbaf3072f8cff330a3c59b2357604cbfc Mon Sep 17 00:00:00 2001
From: svyat-innowise <svyatoslav.yaroshevich@innowise-group.com>
Date: Tue, 21 May 2024 17:14:54 +0200
Subject: [PATCH] zephyr: lvgl_fs: fix bug of memory lvgl leak

---
 modules/lvgl/lvgl_fs.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/modules/lvgl/lvgl_fs.c b/modules/lvgl/lvgl_fs.c
index a05e21748d..aa61493d60 100644
--- a/modules/lvgl/lvgl_fs.c
+++ b/modules/lvgl/lvgl_fs.c
@@ -86,8 +86,10 @@ static void *lvgl_fs_open(struct _lv_fs_drv_t *drv, const char *path, lv_fs_mode
 static lv_fs_res_t lvgl_fs_close(struct _lv_fs_drv_t *drv, void *file)
 {
 	int err;
-
+	
 	err = fs_close((struct fs_file_t *)file);
+	if(file!=NULL)
+	LV_MEM_CUSTOM_FREE(file);
 	return errno_to_lv_fs_res(err);
 }
 
@@ -236,4 +238,4 @@ void lvgl_fs_init(void)
 	fs_drv.dir_close_cb = lvgl_fs_dir_close;
 
 	lv_fs_drv_register(&fs_drv);
-}
+}
\ No newline at end of file
-- 
2.41.0.windows.1


```

**Reply by faxe1008** [reply]
Thanks for raising the issue and investigating this. I opened an issue on the basis of this and submitted a PR. :^)

**Reply by MaprilBear** [reply]
After debugging this memory leak for hours and hours on Zephyr v3.5 (zmk is still suck on it unfortunately), I'm happy to see this has been fixed already! Thanks for saving folks from the black hole that I fell into :P
